numpy/lib/function_base.py

```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
-        lerp_interpolation = lerp_interpolation[()]  # unpack 0d arrays
-    return lerp_interpolation
+def _virtual_index_formula(array_size: np.array,
+                           quantiles: np.array,
+                           alpha: float,
+                           beta: float) -> np.array:
+    """
+   Compute the floating point indexes of an array for the linear
+   interpolation of quantiles.
+   Reference:
+   Hyndman&Fan paper "Sample Quantiles in Statistical Packages",
+   DOI: 10.1080/00031305.1996.10473566
+   """
+    return array_size * quantiles + (
+            alpha + quantiles * (1 - alpha - beta)
+    ) - 1
+
+
+def _get_gamma(type: np.dtype,
+               virtual_indexes: np.array,
+               previous_indexes: np.array,
+               interpolation: QuantileInterpolation) -> np.array:
+    """
+    Compute the gamma (a.k.a 'm' or weight) for the linear interpolation
+    of quantiles.
+    When gamma == 0 the left bound will be picked,
+    When gamma == 1 the right bound will be.
+    """
+    gamma = np.asanyarray(virtual_indexes - previous_indexes)
+    gamma = interpolation.get_gamma_correction(gamma, virtual_indexes)
+    return np.asanyarray(gamma)
 
 
-def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
-                           interpolation='linear', keepdims=False):
-    a = asarray(a)
+def _linear_interpolation_formula(
+        left: np.array, right: np.array, gamma: np.array, out: np.array = None
+) -> np.array:
+    """
+    Compute the linear interpolation weighted by gamma on each point of
+    two same shape array.
+    """
+    # Equivalent to gamma * right + (1 - gamma) * left
+    # see gh-14685
+    diff_right_left = subtract(right, left)
+    result = asanyarray(add(left, diff_right_left * gamma, out=out))
+    result = subtract(right,
+                      diff_right_left * (1 - gamma),
+                      out=result,
+                      where=gamma >= 0.5)
+    return result
 
-    # ufuncs cause 0d array results to decay to scalars (see gh-13105), which
-    # makes them problematic for __setitem__ and attribute access. As a
-    # workaround, we call this on the result of every ufunc on a possibly-0d
-    # array.
-    not_scalar = np.asanyarray
 
-    # prepare a for partitioning
-    if overwrite_input:
-        if axis is None:
-            ap = a.ravel()
-        else:
-            ap = a
-    else:
-        if axis is None:
-            ap = a.flatten()
-        else:
-            ap = a.copy()
+def _get_gamma_mask(shape, default_value, value, where):
+    out = np.full(shape, default_value)
+    out[where] = value
+    return out
 
-    if axis is None:
-        axis = 0
 
+def get_interpolation_method(unknown: np.array) -> QuantileInterpolation:
+    if isinstance(unknown, QuantileInterpolation):
+        return unknown
+    elif isinstance(unknown, str):
+        for inter in QuantileInterpolation:
+            if unknown.lower() in inter.aliases:
```

Comment:
Are you sure we should be calling `lower` here? Did we support mixed case before?